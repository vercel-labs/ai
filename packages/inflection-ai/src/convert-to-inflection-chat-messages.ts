import {
  LanguageModelV1Prompt,
  LanguageModelV1TextPart,
  UnsupportedFunctionalityError,
} from '@ai-sdk/provider';

/**
 * Represents a message in the Inflection AI format
 */
export interface InflectionMessage {
  /** The type of the message sender */
  type: 'Human' | 'AI' | 'Instruction';
  /** The text content of the message */
  text: string;
  /** Optional timestamp in seconds since epoch */
  ts?: number;
}

export type InflectionContext = InflectionMessage[];

/**
 * Converts a Language Model prompt to Inflection AI's context format
 * @param prompt The input prompt in the Language Model format
 * @returns An array of messages in Inflection AI's format
 * @throws {UnsupportedFunctionalityError} If the prompt contains unsupported content types
 */
export function convertToInflectionChatMessages(
  prompt: LanguageModelV1Prompt,
): InflectionContext {
  const context: InflectionContext = [];

  for (const { role, content } of prompt) {
    let text = '';

    if (typeof content === 'string') {
      text = content;
    } else {
      // Combine all text parts into a single string
      for (const part of content) {
        switch (part.type) {
          case 'text': {
            text += part.text;
            break;
          }
          case 'image': {
            throw new UnsupportedFunctionalityError({
              functionality:
                'Image content parts are not supported by Inflection AI at this time',
            });
          }
          case 'file': {
            throw new UnsupportedFunctionalityError({
              functionality:
                'File content parts are not supported by Inflection AI at this time',
            });
          }
          case 'tool-call': {
            throw new UnsupportedFunctionalityError({
              functionality:
                'Tool calls are not supported by Inflection AI at this time',
            });
          }
          case 'tool-result': {
            throw new UnsupportedFunctionalityError({
              functionality:
                'Tool results are not supported by Inflection AI at this time',
            });
          }
          default: {
            const _exhaustiveCheck: never = part;
            throw new Error(
              `Unsupported content part type: ${_exhaustiveCheck}`,
            );
          }
        }
      }
    }

    if (!text) {
      continue; // Skip empty messages
    }

    // Map the role to Inflection's expected type
    switch (role) {
      case 'user': {
        context.push({ type: 'Human', text });
        break;
      }
      case 'assistant': {
        context.push({ type: 'AI', text });
        break;
      }
      case 'system': {
        context.push({ type: 'Instruction', text });
        break;
      }
      case 'tool': {
        // Tool responses are handled as part of the content processing above
        break;
      }
      default: {
        const _exhaustiveCheck: never = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }

  return context;
}
