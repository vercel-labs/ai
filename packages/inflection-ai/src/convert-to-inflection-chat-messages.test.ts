import { expect, describe, it } from 'vitest';
import { UnsupportedFunctionalityError } from '@ai-sdk/provider';
import { convertToInflectionChatMessages } from './convert-to-inflection-chat-messages';

describe('convertToInflectionChatMessages', () => {
  it('should convert basic user and assistant messages', () => {
    const result = convertToInflectionChatMessages([
      {
        role: 'user',
        content: [{ type: 'text' as const, text: 'Hello' }],
      },
      {
        role: 'assistant',
        content: [{ type: 'text' as const, text: 'Hi there!' }],
      },
    ]);

    expect(result).toEqual([
      { type: 'Human', text: 'Hello' },
      { type: 'AI', text: 'Hi there!' },
    ]);
  });

  it('should convert system messages to instruction type', () => {
    const result = convertToInflectionChatMessages([
      {
        role: 'system',
        content: 'Be helpful and concise.',
      },
    ]);

    expect(result).toEqual([
      { type: 'Instruction', text: 'Be helpful and concise.' },
    ]);
  });

  it('should combine multiple text parts', () => {
    const result = convertToInflectionChatMessages([
      {
        role: 'user',
        content: [
          { type: 'text' as const, text: 'Hello ' },
          { type: 'text' as const, text: 'world!' },
        ],
      },
    ]);

    expect(result).toEqual([{ type: 'Human', text: 'Hello world!' }]);
  });

  it('should throw error for image content', () => {
    expect(() =>
      convertToInflectionChatMessages([
        {
          role: 'user',
          content: [
            {
              type: 'image',
              image: new Uint8Array([1, 2, 3]),
              mimeType: 'image/jpeg',
            },
          ],
        },
      ]),
    ).toThrow(UnsupportedFunctionalityError);
  });

  it('should throw error for tool results', () => {
    expect(() =>
      convertToInflectionChatMessages([
        {
          role: 'tool',
          content: [
            {
              type: 'tool-result',
              toolCallId: 'tool-call-id-1',
              toolName: 'tool-1',
              result: { key: 'result-value' },
            },
          ],
        },
      ]),
    ).toThrow(UnsupportedFunctionalityError);
  });

  it('should skip empty messages', () => {
    const result = convertToInflectionChatMessages([
      {
        role: 'user',
        content: [{ type: 'text' as const, text: '' }],
      },
    ]);

    expect(result).toEqual([]);
  });
});

describe('tool calls', () => {
  it('should throw error for tool calls', () => {
    expect(() =>
      convertToInflectionChatMessages([
        {
          role: 'assistant',
          content: [
            {
              type: 'tool-call',
              args: { key: 'arg-value' },
              toolCallId: 'tool-call-id-1',
              toolName: 'tool-1',
            },
          ],
        },
      ]),
    ).toThrow(UnsupportedFunctionalityError);
  });
});

describe('assistant messages', () => {
  it('should add prefix true to trailing assistant messages', () => {
    const result = convertToInflectionChatMessages([
      {
        role: 'user',
        content: [{ type: 'text' as const, text: 'Hello' }],
      },
      {
        role: 'assistant',
        content: [{ type: 'text' as const, text: 'Hello!' }],
      },
    ]);

    expect(result).toMatchSnapshot();
  });
});
