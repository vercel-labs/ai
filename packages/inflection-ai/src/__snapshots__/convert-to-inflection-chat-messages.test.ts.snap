// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assistant messages > should add prefix true to trailing assistant messages 1`] = `
[
  {
    "text": "Hello",
    "type": "Human",
  },
  {
    "text": "Hello!",
    "type": "AI",
  },
]
`;

exports[`tool calls > should stringify arguments to tool calls 1`] = `[]`;
